File: 3-HARD/Python/5-line-sweep-closest-points/proposed-solution/closest_points.py
from typing import List, Tuple
import math
import bisect

class ClosestPoints:
    def __init__(self, points: List[Tuple[int, int]]):
        self.points = sorted(points, key=lambda x: x[0])

    def distance(self, p1: Tuple[int, int], p2: Tuple[int, int]) -> float:
        return math.sqrt((p1[0] - p2[0])**2 + (p1[1] - p2[1])**2)

    def find_closest_pair(self) -> Tuple[Tuple[int, int], Tuple[int, int], float]:
        active = []
        best_pair = (None, None)
        best_dist = float('inf')
        left = 0

        for point in self.points:
            while active and point[0] - self.points[left][0] > best_dist:
                bisect.insort_left(active, self.points[left][1])
                left += 1

            y_lower = point[1] - best_dist
            y_upper = point[1] + best_dist
            i = bisect.bisect_left(active, y_lower)
            j = bisect.bisect_right(active, y_upper)

            for y in active[i:j]:
                candidate = (point[0], y)
                dist = self.distance(point, candidate)
                if dist < best_dist:
                    best_dist = dist
                    best_pair = (point, candidate)

            bisect.insort(active, point[1])

        return best_pair[0], best_pair[1], best_dist

File: 3-HARD/JavaScript/5-line-sweep-closest-points/proposed-solution/ClosestPoints.js
// ClosestPoints.js - Line Sweep Algorithm to find closest pair of points in JavaScript

class ClosestPoints {
  constructor(points) {
    this.points = points.slice().sort((a, b) => a[0] - b[0]);
  }

  distance(p1, p2) {
    return Math.sqrt((p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2);
  }

  findClosestPair() {
    const active = [];
    let bestPair = [null, null];
    let bestDist = Infinity;
    let left = 0;

    const bisectLeft = (arr, x) => {
      let low = 0,
        high = arr.length;
      while (low < high) {
        const mid = Math.floor((low + high) / 2);
        if (arr[mid] < x) low = mid + 1;
        else high = mid;
      }
      return low;
    };

    const bisectRight = (arr, x) => {
      let low = 0,
        high = arr.length;
      while (low < high) {
        const mid = Math.floor((low + high) / 2);
        if (arr[mid] <= x) low = mid + 1;
        else high = mid;
      }
      return low;
    };

    for (const point of this.points) {
      while (active.length > 0 && point[0] - this.points[left][0] > bestDist) {
        // Remove points[left][1] from active using binary search
        const idx = bisectLeft(active, this.points[left][1]);
        if (idx < active.length && active[idx] === this.points[left][1]) {
          active.splice(idx, 1);
        }
        left++;
      }

      const yLower = point[1] - bestDist;
      const yUpper = point[1] + bestDist;
      const i = bisectLeft(active, yLower);
      const j = bisectRight(active, yUpper);

      for (let k = i; k < j; k++) {
        const candidate = [point[0], active[k]];
        const dist = this.distance(point, candidate);
        if (dist < bestDist) {
          bestDist = dist;
          bestPair = [point, candidate];
        }
      }

      // Insert point[1] into active in sorted order
      const pos = bisectLeft(active, point[1]);
      active.splice(pos, 0, point[1]);
    }

    return { pair: bestPair, distance: bestDist };
  }
}

export default ClosestPoints;
